### DISCUSSION TITLE
Error in OpenAPI schema generation with plain serializer when upgrading to pydantic 2.8.0

### DISCUSSION THREAD

**Discussion by rossmacarthur** [fastapi/fastapi#11789] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from decimal import ROUND_HALF_UP, Decimal
from typing import Annotated, Optional

import fastapi
from pydantic import BaseModel, PlainSerializer


def cents_to_dollars(v: Decimal) -> str:
    v = v * Decimal("0.01")
    v = v.quantize(Decimal("0.00000000"), rounding=ROUND_HALF_UP)
    return format(v, "f")


serializer = PlainSerializer(cents_to_dollars, when_used="json-unless-none")


class Example(BaseModel):
    a: Annotated[Decimal, serializer]  # works fine
    b: Annotated[Optional[Decimal], serializer]  # works fine
    c: Annotated[Optional[Decimal], serializer] = None  # does not work


app = fastapi.FastAPI()


@app.get("/")
def root() -> Example: ...


print(app.openapi())
```

### Description

The above code works fine on pydantic 2.7.4 but errors on pydantic 2.8.0

### Operating System

macOS

### FastAPI Version

0.111.0

### Pydantic Version

2.8.0

### Python Version

3.11.7

**Comment by danielqba** [comment]
Had the same issue around here, had to pin the pydantic version to 2.7.4

**Comment by ceb10n** [comment]
Apparently, for some reason that I haven't investigated yet, you need to add `SerializationInfo` to your serializer:

```python
def cents_to_dollars(v: Decimal, info: SerializationInfo) -> str:
    v = v * Decimal("0.01")
    v = v.quantize(Decimal("0.00000000"), rounding=ROUND_HALF_UP)
    return format(v, "f")
```

I just saw it in the the [Pydantic v2.8  announcement](https://pydantic.dev/articles/pydantic-v2-8-release) and tryied with your example.

Since it is a problem with your function beeing called with `v = None` when `SerializationInfo` is not present, you can validate if `v` is not `None` too:

```python
def cents_to_dollars(v: Decimal) -> str:
    if v:
        v = v * Decimal("0.01")
        v = v.quantize(Decimal("0.00000000"), rounding=ROUND_HALF_UP)
        return format(v, "f")
```

**Reply by rossmacarthur** [reply]
I think FastAPI is not considering this parameter: `when_used="json-unless-none"`,  because when set the serializer is not actually called when the value is `None` but FastAPI is trying to call it on the default value `None` in this case.
